### PR TITLE
Adds the namespace migrator

### PR DESCRIPTION
**Pull request by jathak**
This allows you to change namespaces for `@use` rules by matching regular expressions on the existing namespace or on the rule URL.

The migrator also has special handling for `@use` rules without member references which can be configured to run only when they conflict with a referenced `@use` rule (default), for all unreferenced rules, or for none of them.

The renaming itself is specified through a small DSL. The basics of this are documented in the changelog and the help text, but I'll add proper documentation to the website before merging this and releasing 1.3.0.

The tests under `test/migrators/namespace` focus on conflict resolution and the actual patching of source files, while the DSL itself is mostly tested in the unit tests in `test/renamer_test.dart`.

### CODE REVIEW

File: lib/src/migrators/namespace.dart [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+      ImportCache importCache, Stylesheet stylesheet, Importer importer) {
+    Renamer<UseRule> renamer;
+    try {
+      renamer = Renamer(argResults['rename'].join('\n'), 'namespace', {
+        'namespace': (rule) => rule.namespace,
+        'url': (rule) => rule.url.toString()
+      });
+    } on FormatException catch (e) {
+      throw MigrationException(e.message);
+    }
```
Comment (nex3) [inline lib/src/migrators/namespace.dart:68]: Also catch a `SourceSpanFormatException` and convert it to a `MigrationSourceSpanException`.

File: lib/src/migrators/namespace.dart [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    var oldUsedNamespaces = _usedNamespaces;
+    _spansByNamespace = {};
+    _usedNamespaces = {};
+    super.visitStylesheet(node);
+    _spansByNamespace = oldSpansByNamespace;
+    _usedNamespaces = oldUsedNamespaces;
+  }
+
+  @override
+  void beforePatch(Stylesheet node) {
```
Comment (nex3) [inline lib/src/migrators/namespace.dart:105]: Breaking this up with some empty lines between blocks would help readability.

File: lib/src/migrators/namespace.dart [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    // Pass each `@use` rule through
```
Comment (nex3) [inline lib/src/migrators/namespace.dart:106]: Through what?

File: lib/src/migrators/namespace.dart [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+  }
+}
+
+class _NamespaceMigrationVisitor extends MigrationVisitor {
+  final Renamer<UseRule> renamer;
+  final UnreferencedFlag unreferencedFlag;
+  final bool forceRename;
+
+  Map<String, Set<FileSpan>> _spansByNamespace;
+  Set<String> _usedNamespaces;
```
Comment (nex3) [inline lib/src/migrators/namespace.dart:87]: Document these.

File: lib/src/migrators/namespace.dart [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+          if (!_spansByNamespace.containsKey(rule.namespace)) {
+            unreferencedRules.add(rule);
+            rules.remove(rule);
+          }
+        }
+      }
+      if (rules.length == 1) {
+        _patchNamespace(rules.first, newNamespace);
+        continue;
+      }
```
Comment (nex3) [inline lib/src/migrators/namespace.dart:138]: Dead code?

File: lib/src/migrators/namespace.dart [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+      if (!forceRename) {
+        throw MigrationException(
+            'Rename failed. ${rules.length} rules would use namespace '
+            '"$newNamespace".\n'
+            'Run with --force to rename with numerical suffixes.');
```
Comment (nex3) [inline lib/src/migrators/namespace.dart:143]: This should include the spans of the rules in question. You can hook into `MultiSpanSassException` to make that easy.

File: lib/src/migrators/namespace.dart [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+      _patchNamespace(rule, namespace);
+    }
+  }
+
+  /// Patch [rule] and all references to it with [newNamespace].
+  void _patchNamespace(UseRule rule, String newNamespace) {
+    _usedNamespaces.add(newNamespace);
+    if (rule.namespace == newNamespace) return;
+    var asClause =
+        RegExp('\\sas\\s(${rule.namespace})').firstMatch(rule.span.text);
```
Comment (nex3) [inline lib/src/migrators/namespace.dart:171]: It's possible that there's multiple characters of whitespace between `as` and the namespace. It's even vaguely possible that there's no whitespace before `as`, although I'm not sure we really need to worry about that.

If it's helpful, you could add `UseRule.urlSpan`, which would make this and the replacement logic below much more reliable.

File: lib/src/migrators/namespace.dart [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    if (asClause == null) {
+      // Add an `as` clause to a rule that previously lacked one.
+      var end = RegExp(r"""@use\s("|').*\1""").firstMatch(rule.span.text).end;
```
Comment (nex3) [inline lib/src/migrators/namespace.dart:174]:
```suggestion
      var end = RegExp(r"""@use\s("|').*?\1""").firstMatch(rule.span.text).end;
```

If the `.*` here is greedy, it would match all the way into a `with` clause that contained quoted string arguments, which is definitely not what we want.

File: test/renamer_test.dart [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+      });
+    });
+
+    group('escaping', () {
+      test(r'spaces', () {
+        var renamer = Renamer.simple(r'a\ b to x\ y');
+        expect(renamer.rename('a b'), equals('x y'));
+      });
+
+      test(r'normal delimiters', () {
```
Comment (nex3) [inline test/renamer_test.dart:84]: Also test a delimiter at the beginning of the source pattern and escaped backslashes, including things like `foo to bar\\;` and `foo to bar\\\;`.

File: test/renamer_test.dart [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+            Renamer.map(r'url .*/(\w+)/lib/mixins to \1', ['namespace', 'url']);
+        expect(
+            renamer.rename(
+                {'namespace': 'mixins', 'url': 'path/button/lib/mixins'}),
+            equals('button'));
+      });
+    });
+  });
+
+  group('multiple statements', () {
```
Comment (nex3) [inline test/renamer_test.dart:116]: Test empty initial statements. In particular, it's likely that users writing shell scripts will write empty initial newlines.

File: test/renamer_test.dart [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+
+    test('only first matching statement is applied', () {
+      var renamer = Renamer.simple('/.*/all/;/old/wrong/;/all/wrong/');
+      expect(renamer.rename('old'), equals('all'));
+      expect(renamer.rename('wrong'), equals('all'));
+      expect(renamer.rename('all'), equals('all'));
+    });
+  });
+
+  group('invalid syntax', () {
```
Comment (nex3) [inline test/renamer_test.dart:149]: Also:

* Empty string
* Three clauses but the middle one isn't `to`
* Four clauses but the first one isn't a known key

File: test/utils.dart [hunk 12]
```diff
@@ -132,7 +132,8 @@ class _HrxTestFiles {
             "given test.";
       }
     } else if (filename == "arguments") {
-      arguments = contents.trim().split(" ");
+      // TODO(jathak): Handle spaces with quotes here.
```
Comment (nex3) [inline test/utils.dart:135]: Not for the first time, I really wish Dart had a shellparse package. In this case, I think you should be able to make do with scanning for instances of `"[^"]+"|[^"\s][^\s]*`

**Review by nex3 — CHANGES_REQUESTED** [review]

File: test/renamer_test.dart (hunk 10 above)
Comment (jathak) [reply test/renamer_test.dart:96]: Done. The updated parser should handle any amount of spaces, newlines, and semicolons at any point before or after a statement.

File: test/renamer_test.dart (hunk 11 above)
Comment (jathak) [reply test/renamer_test.dart]: Done. I decided to make a renamer with no statements not actually an error (it just will always return null), so that if someone wants to run this migrator just for `--unreferencable=all` it should still work.

File: test/utils.dart (hunk 12 above)
Comment (jathak) [reply test/utils.dart]: Done.

**Review by jathak — COMMENTED** [review]
Decided to remove the sed-style syntax after finding some edge cases where it's ambiguous which syntax a statement should be parsed as.

I rewrote the parser accordingly using StringScanner.

File: lib/src/renamer.dart [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+
+  /// Reads the next statement (and the trailing delimiter, if any) from
+  /// [scanner].
+  static _Statement<T> _readStatement<T>(
+      StringScanner scanner, Map<String, String Function(T input)> keys) {
+    var start = scanner.position;
+    FormatException lastException;
+    // Tries each key in succession until one is successfully returned.
+    for (var entry in keys.entries) {
+      try {
```
Comment (nex3) [inline lib/src/renamer.dart:100]: If I'm understanding correctly, the reason you have a `try`/`catch` here is to handle the case where a user uses the default key is used for a pattern that's also the same as a key name (e.g. `url to my-url`), right? That's a complex enough case that it's worth documenting in a comment (and testing if you aren't already).

For reference, catching exceptions like this to handle control flow is quite slow. This code is not performance-critical at all so it probably doesn't matter, but if you wanted to make something like this more efficient you could potentially do so by doing a "lexing" pass where you handle whitespace separators and escape sequences and return a list of strings in the next statement, and then a "parsing" phase where you can make decisions based on how many strings there are.

File: lib/src/renamer.dart [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+        scanner.position = start;
```
Comment (nex3) [inline lib/src/renamer.dart:101]: I'd do this in the `catch` clause, so that it's clear that the reason the position is in a weird state is because you're handling an exception.

File: lib/src/renamer.dart [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+    var output = _readOutput(scanner);
+    return _Statement(keyFunction, matcher, output);
+  }
+
+  /// Reads a matcher clause and its trailing space from [scanner].
+  static RegExp _readMatcher(StringScanner scanner) {
+    int quote;
+    if ({$single_quote, $double_quote}.contains(scanner.peekChar())) {
+      quote = scanner.readChar();
+    }
```
Comment (nex3) [inline lib/src/renamer.dart:135]: I think people who are used to JS or shell idioms would get pretty confused using quotes to delimit a regular expression. I'd just stick with allowing spaces to be backslash-escaped, especially since in the use-cases we know of we expect it to be very rare that someone would want to write a space in the input or output. As a bonus, that'll also make this logic simpler :smile:.

**Review by nex3 — CHANGES_REQUESTED** [review]
Looking good!

File: CHANGELOG.md [hunk 16]
```diff
@@ -1,3 +1,43 @@
...
+  By default, if the renaming results in a conflict between multiple `@use`
+  rules, the migration will fail, but you can force it to resolve conflicts with
+  numerical suffixes by passing `--force`.
+
+  This migrator will also check for `@use` rules without referenced members. If
+  any of these results has a conflict, it will be resolved by namespacing the
+  unreferenced rule with `_unreferenced#` instead of failing. You can disable
+  this behavior with `--unreferenced=none` or have the migrator rename *all*
+  unreferenced `@use` rules with this pattern (instead of just conflicting ones)
+  by passing `--unreferenced=all`.
```
[the remark this answers is not in the document]
Comment (jathak) [reply CHANGELOG.md]: Removed

File: lib/src/renamer.dart [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+    return Renamer(code, {for (var key in keys) key: (input) => input[key]});
+  }
+
+  /// Creates a Renamer based on [code] and a map from keys to functions that
+  /// take an input and return a string.
+  ///
+  /// All keys must consist of only lowercase letters, underscores, and hyphens.
+  ///
+  /// If provided, [sourceUrl] will appear in parsing errors. It can be
+  /// a [String] or a [Uri].
```
[the remark this answers is not in the document]
Comment (jathak) [reply lib/src/renamer.dart:80]: Added. This was necessary to provide some useful context to any exceptions that StringScanner throws (I pass `--rename` as the source "URL" so that any parsing exceptions there indicate that it's from the rename option, and not from an actual Sass file.

File: lib/src/renamer.dart (hunk 13 above)
Comment (jathak) [reply lib/src/renamer.dart:107]: Added a comment and a couple of tests.

I considered that kind of two-pass parse, but the matcher and output clauses have slightly different behaviors for escape sequences (since the matcher clause needs to leave most of them intact for the regex parser to handle, but the output clause handles them all). Because of that, I figured the performance cost of the exceptions was worth the simpler implementation given that only the parsing of the renamer scripts would be slowed, not the actual execution.

File: lib/src/renamer.dart (hunk 14 above)
Comment (jathak) [reply lib/src/renamer.dart]: Done

File: lib/src/renamer.dart (hunk 15 above)
Comment (jathak) [reply lib/src/renamer.dart]: Removed

**Review by nex3 — APPROVED** [review]
